arangod/Cluster/ClusterInfo.cpp

```diff
@@ -2465,76 +2491,145 @@ int ClusterInfo::ensureIndexCoordinator(
+          // Will wait, if we have not gone over time
+          if (sleepFor <= 2500) {
+            sleepFor*=2;
+          }
+          std::this_thread::sleep_for(std::chrono::milliseconds());
+          continue;
         }
-      }
-    }
 
-    if (planValue == nullptr) {
-      // hmm :S both empty :S did somebody else clean up? :S
-      // should not happen?
-      return errorCode;
-    }
+        if (overTime) {
+          current.reset(
+            TRI_ERROR_CLUSTER_TIMEOUT,
+            "Timed out while waiting for indexes to get ready on db servers");
+        }
+
+        errorCode = TRI_ERROR_CLUSTER_TIMEOUT;
+        setErrormsg(errorCode, errorMsg);
+        
+      }
+      
+      std::vector<AgencyOperation> opers;
+      if (current.ok()) {
+        opers.push_back(
+          AgencyOperation(
+            indexPath, AgencyValueOperationType::REPLACE,
+            newPlanIndex.slice(), oldPlanIndex.slice()));
+      } else {
+        opers.push_back(
+          AgencyOperation(
+            indexPath, AgencyValueOperationType::ERASE, oldPlanIndex.slice()));
+      }
+      opers.push_back(
+        AgencyOperation(
+          "Plan/Version", AgencySimpleOperationType::INCREMENT_OP));
+      AgencyWriteTransaction trx(opers);
+      while (true) {
+        AgencyCommResult update =
+          _agency.sendTransactionWithFailover(trx, 0.0);
+        if (update.successful()) {
+          loadPlan();
+          return current.errorNumber();
+        }
+      }
 
-    std::string const planIndexesKey = "Plan/Collections/" + databaseName + "/" + collectionID + "/indexes";
-    std::vector<AgencyOperation> operations;
-    std::vector<AgencyPrecondition> preconditions;
-    if (planValue) {
-      AgencyOperation planEraser(planIndexesKey, AgencyValueOperationType::ERASE, planValue->slice());
-      TRI_ASSERT(oldPlanIndexes);
-      AgencyPrecondition planPrecondition(planIndexesKey, AgencyPrecondition::Type::VALUE, oldPlanIndexes->slice());
-      operations.push_back(planEraser);
-      operations.push_back(AgencyOperation("Plan/Version", AgencySimpleOperationType::INCREMENT_OP));
-      preconditions.push_back(planPrecondition);
+    } else {
+      // That's odd isBuilding key has gone already. Supervision was quicker?
+      loadPlan();
+      return TRI_ERROR_NO_ERROR;
     }
+    
+  }
 
-    AgencyWriteTransaction trx(operations, preconditions);
-    AgencyCommResult eraseResult = _agency.sendTransactionWithFailover(trx, 0.0);
+  // At this time the Iindex creation has failed and we want to  roll back
+  // the plan entry
+  VPackBuilder oldPlanSlice;
+  AgencyWriteTransaction trx(
+    std::vector<AgencyOperation>
+    { AgencyOperation(
+           indexPath, AgencyValueOperationType::ERASE, oldPlanIndex.slice()),
+        AgencyOperation(
+        "Plan/Version", AgencySimpleOperationType::INCREMENT_OP)});
 
-    if (eraseResult.successful()) {
+  setErrormsg(errorCode, errorMsg);
+  
+  while (true) {
```

Comment:
This loop can turn forever, if the ERASE does not work for some reason.

Reply:
fixed